R/functions.R

```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
+#' @param housestruct House structure vector that gives number of kids in each household.
+#'
+#' @return an N x 3 matrix (N = # of observations (kids)), the first column identifies which households are treated with a "1" for the first kid in a treated household, the second and third columns are the ones that matter, they are the treatment and exposure vectors at the observation (kid) level.
+#' @export
+out_Z_household = function(N,K,treatment_list,housestruct){
+  which_house = sort(treatment_list$which_house)
+  theord = order(treatment_list$which_house)
+  which_kid = treatment_list$which_kid[theord]
+  
+  Z = matrix(0,nrow=N,ncol=3)
+  colnames(Z) = c('houseind','treat','exp')
+  houseind = cumsum(housestruct)-housestruct[1]+1
+  Z[houseind,1] = 1 # separates out the different houses
+  treatedind = houseind[which_house] + which_kid - 1
+  Z[treatedind,2] = 1
+  lind = houseind[which_house] 
+  hind = lind + housestruct[which_house] - 1
+  expind = c(apply(as.matrix(1:length(which_house)),1,out_exp_ind,lind=lind,hind=hind))
+  Z[expind,3] = 1
+  Z
+}
+
+#' Returns matrix \code{Z} of dimension (number of units x number of randomizations, i.e. assignments.)
+#'
+#' @param N The number of observations (kids).
+#' @param K The number of houses.
+#' @param equal A binary parameter that determines the exact house structure. If equal=TRUE, every household has equal number of kids. If equal=FALSE, every household samples the number of kids at random.
+#' @param numrand The number of randomizations
+#' 
+#' @return A matrix compounds treated units as well as exposure functions. 0 represents a pure control unit. 1 represents a spillover unit. And 2 represents a treated unit.
+#' @export
+out_Zprime = function(N,K,equal=T,numrand){
+  housestruct = out_house_structure(N,K,equal)
+  Zprime_mat = matrix(0,nrow=N,ncol=numrand)
+  for(nn in 1:numrand){
+    treatment_list = out_treat_household(housestruct,K1=K/2)  
+    Zp = out_Z_household(N,K,treatment_list,housestruct)
+    Zp_compact = rowSums(Zp[,2:3])
+    Zprime_mat[,nn] = Zp_compact
+  }
+  Zprime_mat
+}
+
+#' Simulates the outcome vector based on the treatment assignment vector. The specific data generating procedure is adopted from Basse & Feller.
+#'
+#' @param N The number of observations (kids).
+#' @param K The number of houses.
+#' @param Zobs Observed treatment assignment vector.
+#' @param tau_main The main effect.
+#' @param sig_c Standard error on the causal effect.
+#' @param sig_y Standard error on the outcome vector.
+#' @param taus Spillover effect.
+#' @param taup Primary effect.
+#' @param mu00 The effect on pure control units.
+#' @param equal A binary parameter that determines the exact house structure. If equal=TRUE, every household has equal number of kids. If equal=FALSE, every household samples the number of kids at random.
+#' 
+#' @return An outcome vector.
+#' @export
+out_bassefeller = function(N, K, Zobs, tau_main,
```

Comment:
maybe include the year of the basse / feller paper when citing, I think it's 2017?  Also, 

(1) You can remove the commented out lines of code in the function below if they are unnecessary
(2) Have you run the example to make sure everything works?

Reply:
Hi David, I have run the example when there is a positive spillover effect and the code returns a p value = 0, which from my prospective, rejects the null hypothesis of no spillover effect. I have also checked the validity under the null hypothesis and it's a valid test.

Reply:
The arxiv version of Basse, Feller's paper is in 2017, but the published version on JASA is in 2018. As a result, I add the year of the paper as 2018.